controllers/kubeadmconfig_controller.go

```diff
@@ -267,10 +267,15 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 	}
 	if certificates != nil {
 		hashes, err := certs.CertificateHashes(certificates.ClusterCA.Cert)
-		if err == nil {
-			config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{
-				CACertHashes: hashes,
+		if err != nil {
+			log.Error(err, "Unable to generate certificate hashes")
+			return ctrl.Result{}, err
+		}
+		if hashes != nil {
+			if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
```

Comment:
IMO this piece of code should be moved into `reconcileDiscovery `, so all the logics impaction `JoinConfiguration.Discovery` are in the same place.
Nb. the bit of code for creating `BootstrapToken` if nil is already implemented there as well

Reply:
I don't think it makes sense to add this check to `reconcileDiscovery` unless you are also suggesting to add certificates into the `reconcileDiscovery` method. We look up certificates to see if we have a CACertHash to use. If we do then `reconcileDiscovery` does something different than if it didn't have a CACertHash. Therefore `reconcileDiscovery` depends on looking up certificates. Either certificates becomes part of `reconcileDiscovery` or we will have to live with not all defaults ending up in `reconcileDiscovery`

If we want to refactor it that's fine but I think that's out of scope for this PR.

The order is a little bit weird because we call `reconcileDiscovery right after we figure out our Certificates